python/ray/data/read_api.py

```diff
@@ -320,6 +323,86 @@ def read_datasource(
+    pipelines: List[List[Dict]],
+    schema: Schema,
+    **kwargs,
+) -> Dataset[ArrowRow]:
+    """Create an Arrow dataset from MongoDB for the given pipelines.
+
+    A MongoDB is described by three elements: URI, database and collection.
+
+    The URI points to an MongoDB instance, for example, a local instance would be
+    "mongodb://localhost:27017", or a hosted instance would be
+    "mongodb+srv://USERNAME:PASSWORD@CLUSTERID.azure.mongodb.net/sample_weatherdata?retryWrites=true&w=majority".
+    For more details, see
+    https://www.mongodb.com/docs/manual/reference/connection-string/.
+
+    The database is similar to the database concept in SQL, and the collection is
+    similar to the table concept in a SQL database.
+
+    To read the MongoDB in parallel, users should provide a list of MongoDB
+    pipelines, with each corresponding to a block to be created for Dataset. Those
+    pipelines are usually formulated as disjoint range queries over a specific field (
+    i.e. partition field). Each pipeline is composed of a series of MongoDB queries,
+    for example, the following is a pipeline which contains 3 steps:
+        pipeline = [
+            {
+                "$match": {"partition_field": {"$gte": 0, "$lt": 100}}
+            }, {
+                "$sort": "sort_field"
+            }, {
+                "$limit": 10
+            }
+        ]
+    Note that a pipeline can contain just a single query.
+
+    Implementation wise, we will use pymongo to connect to MongoDB, and use pymongoarrow
+    to convert MongoDB documents to/from Arrow format, which is a supported block format
+    in Dataset.
+
+    Examples:
+        >>> import ray
+        >>> from pymongoarrow.api import Schema
+        >>> ds = ray.data.read_mongo( # doctest: +SKIP
+        ...     uri="mongodb://username:password@mongodb0.example.com:27017/?authSource=admin", # noqa: E501
+        ...     database="my_db",
+        ...     collection="my_collection",
+        ...     pipelines=[[{"$match": {"col2": {"$gte": 0, "$lt": 100}}}]],
+        ...     schema=Schema({"col1": pa.string(), "col2": pa.int64()}),
+        ... )
+
+    Args:
+        uri: The URI of the source MongoDB where the dataset will be
+            read from. For the URI format, see details in
+            https://www.mongodb.com/docs/manual/reference/connection-string/.
+        database: The name of the database hosted in the MongoDB.
+        collection: The name of the collection in the database.
+        pipelines: A list of pipelines that are used to create blocks, with each
+            corresponding to a block. Each pipeline is a list of MongoDB queries (
+            typed as List[Dict]).
+        schema: The schema used to read the collection.
+
+    Returns:
+        Dataset holding Arrow records read with give pipelines from the specified
+        MongoDB collection.
+    """
+    return read_datasource(
+        MongoDatasource(),
+        parallelism=len(pipelines),
```

Comment:
it seems not the right parallelism for me. A MongoDB collection (table) can have multiple documents (files). The whole collection might be too slow to be read in one task if it has a lot of documents. So if user have zero or one pipeline, we basically just launch one task to do the job.

E.g. MongoDB data source for Spark would only support one pipeline (not multiple), and it shards the documents across the tasks to achieve parallel processing - https://github.com/mongodb/mongo-spark/blob/main/src/main/java/com/mongodb/spark/sql/connector/read/partitioner/SamplePartitioner.java#L108-L141 .

Reply:
Yup, we should be like the rest of the datasources and have `parallelism` as a separate parameter. We can insert a repartition op if the number of pipelines is less than the parallelism.

Reply:
@c21 "A list of pipelines that are used to create blocks, with each pipeline corresponding to a block", why is not right?
@ericl Yes, that's an item to do here.